torch/lib/THD/base/data_channels/DataChannelGloo.hpp

```diff
@@ -44,32 +223,29 @@ const ::gloo::ReductionFunction<T>* THDToGlooReduceOp(THDReduceOp op) {
   }
 }
 
-template<DataOperation D, typename T>
-struct algorithm_spec;
-
 template<typename T>
 struct algorithm_spec<DataOperation::ALL_GATHER, T> {
   static GlooCache::key_type key(
     THDGroup group_id, std::size_t input_bytes, std::size_t output_bytes,
     std::size_t unused_count
   ) {
     return std::make_tuple(DataOperation::ALL_GATHER, group_id,
-                           input_bytes, output_bytes, 0);
+                           input_bytes, output_bytes, THDReduceMIN, 0);
```

Comment:
Ugh I don't really like storing all kinds of algorithms in a single map, because you and up with so many unused elements in the keys 😕 I guess we can leave it like that, but at least define some special values:
```cpp
const THDReduceOp UNUSED_OP = THDReduceMIN;
const rank_type UNUSED_RANK = -1;
```